modules/library/render/src/main/java/org/geotools/renderer/label/LabelCacheImpl.java

```diff
@@ -1556,15 +1560,50 @@ private boolean paintPolygonLabel(
             textStyle.setAnchorX(0.5);
             textStyle.setAnchorY(0.5);
         }
-        AffineTransform tx = new AffineTransform(tempTransform);
-        if (paintPolygonLabelInternal(
-                painter, tx, displayArea, glyphs, labelItem, pg, centroid, textStyle)) return true;
+
+        AffineTransform tx = null;
+        boolean allowShrinking =
+                labelItem.getFontShrinkSizeMin() > DEFAULT_FONT_SHRINK_SIZE_MIN
+                        && labelItem.getFontShrinkSizeMin() < textStyle.getFont().getSize();
+        int shrinkSize =
+                allowShrinking ? labelItem.getFontShrinkSizeMin() : textStyle.getFont().getSize();
+        int textSize = textStyle.getFont().getSize();
+        // if shrinking is allowed then try to paint polygon label. If no success reduce font size
+        // by 1 unit and retry until fontShrinkSize is reached.
+        while (textSize >= shrinkSize) {
+            tx = new AffineTransform(tempTransform);
+            LabelCacheItem labelItem2 = painter.getLabel();
+            TextStyle2DExt textStyle2 = new TextStyle2DExt(labelItem2);
+            if (labelItem2.getMaxDisplacement() > 0) {
```

Comment:
What if the style editor actually wanted a given displacement and anchor? I'd suggest to respect the anchoring....

Reply:
actually this is the was the bahavior so far (check also if statement at lines 1557-1562) that is if a max displacement option is set then anchor point seems to be ignored during label checking and assumed by default as point 0.5, 0.5. I did not touch this behavior at all.
Do you want to change this? Moreover should this be done as part of this PR? It seems that this is a different issue that should be addressed separetely.

Reply:
Oh ok, I see it now